charmcraft/config.py

```diff
@@ -0,0 +1,152 @@
+    if not isinstance(value, field.type):
+        key = _get_key(config, field)
+        raise CommandError(
+            "Bad charmcraft.yaml content; field '{}' must be a {}: got {!r}."
+            .format(key, field.type.__name__, value))
+
+
+def _check_url(config, field, value):
+    """Check that the URL has at least scheme and net location."""
+    url = urlparse(value)
+    if not url.scheme or not url.netloc:
+        key = _get_key(config, field)
+        raise CommandError(
+            "Bad charmcraft.yaml content; field '{}' must be a full URL "
+            "(e.g. 'https://some.server.com'): got {!r}.".format(key, value))
+
+
+def _check_typefield(config, field, value):
+    """Check that the value of 'type' field is valid."""
+    # None is allowed until we make mandatory this configuration field
+    if value not in ('charm', 'bundle', None):
+        key = _get_key(config, field)
+        raise CommandError(
+            "Bad charmcraft.yaml content; field '{}' (if present) must value 'charm' or 'bundle': "
+            "got {!r}.".format(key, value))
+
+
+@attr.s(kw_only=True, frozen=True)
+class _CharmhubConfig:
+    section = 'charmhub'
+
+    api_url = attr.ib(
+        type=str, default='https://api.staging.charmhub.io',
+        validator=[_check_url, _check_type])
+    storage_url = attr.ib(
+        type=str, default='https://storage.staging.snapcraftcontent.com',
+        validator=[_check_url, _check_type])
+
+    @classmethod
+    def from_dict(cls, source=None):
+        """Build from a raw dict."""
+        if source is None:
+            return cls()
+
+        if not isinstance(source, dict):
+            raise CommandError(
+                "Bad charmcraft.yaml content; the 'charmhub' field must be a dict: got {!r}."
+                .format(type(source).__name__))
+        return cls(**source)
+
+
+class _BasicPrime(tuple):
+    """Hold the list of files to include, specified under parts/bundle/prime configs.
+
+    This is a intermediate structure until we have the full Lifecycle in place.
+    """
+    @classmethod
+    def from_dict(cls, parts):
+        """Build from a dicts sequence."""
+        if not isinstance(parts, dict):
+            raise CommandError(
+                "Bad charmcraft.yaml content; the 'parts' field must be a dict: got {!r}."
+                .format(type(parts).__name__))
+
+        bundle = parts.get('bundle', {})
+        if not isinstance(bundle, dict):
+            raise CommandError(
+                "Bad charmcraft.yaml content; the 'parts.bundle' field must be a dict: got {!r}."
+                .format(type(bundle).__name__))
+
+        prime = bundle.get('prime', [])
+        if not isinstance(prime, list):
+            raise CommandError(
+                "Bad charmcraft.yaml content; the 'parts.bundle.prime' field must be a list: "
+                "got {!r}.".format(type(prime).__name__))
+
+        # validate that all are relative
```

Comment:
Should this not handle:

```
$ ls prime
foo/ bar foobar
```

- Only Include foo and everything inside it
```yaml
prime:
    - foo
```

- Exclude foo and everything inside it (keeping bar and foobar)
```yaml
prime:
    - -foo
```

Reply:
If so, spec[0] would need some extra logic

Reply:
I don't like this "everything but this" rule, it suffers the same problem than "including everything by default": today you exclude `foo` and get `bar` and `foobar`, tomorrow you also add some confidential keys file and it will be included in the charm.

That said, note that this "basic prime" is a simple intermediate thing that will lead us towards the "full lifecycle capabilities", it doesn't need to cover all the cases.